### DISCUSSION TITLE
How should we handle hybrid coordinates in iHAMOCC?

### DISCUSSION THREAD

**Discussion by TomasTorsvik** [NorESMhub/BLOM#152] [General]
@JorgSchwinger @jmaerz @matsbn 

We should make some changes in iHAMOCC to accommodate the change to hybrid coordinates. At the moment it is assumed that the top two layers represent the surface mixed layer (layer 1: top 10m water column, layer 2: rest of the mixed layer). Special treatment of the surface mixed layer mostly reference the `INTEGER, PARAMETER :: kmle = 2` index (i.e. `do k=1,kmle` etc.) that is defined in the `mo_vgrid` module. The relevant modules/subroutines where this comes into play (that I have found):

##### Preformed tracers in the mixed layer
These tracers are initialized in `beleg_vars` and updated each time step in `preftrc`.
```
ocetra(i,j,k,iprefo2) =ocetra(i,j,k,ioxygen)
ocetra(i,j,k,iprefpo4)=ocetra(i,j,k,iphosph)
ocetra(i,j,k,iprefalk)=ocetra(i,j,k,ialkali)
ocetra(i,j,k,iprefdic)=ocetra(i,j,k,isco212)
```

##### Saturation DIC concentration in mixed layer
`carchm` does some calculations specifically for the surface layer (i.e. `if (k.eq.1) then` etc.), and at one point calculates the saturation DIC concentration in the mixed layer, without reference to `kmle`
```
ocetra(i,j,k,  idicsat)=tc_sat * rrho ! convert mol/kg to kmol/m^3
ocetra(i,j,k+1,idicsat)=tc_sat * rrho ! k+1 = the rest of the mixed layer
```

##### Riverine tracer inputs
Riverine tracer inputs are distributed over the surface mixed layer, for instance
```
ocetra(i,j,1:kmle,iano3) = ocetra(i,j,1:kmle,iano3) + rivin(i,j,irdin)*fdt/volij
```

##### Nitrogen fixation
The `cyano` subroutine is limited by `kmle` to update tracers only in the surface mixed layer.

##### Particle aggregation
This only comes into play if the `AGG` pre-processor flag is defined. In this case the following parameters from `beleg_parm` are used
```
shear = 43200.   !shear in the mixed layer,   0.5  d-1
fsh = 0.163 * shear *dtb
nmldmin = 0.1    ! minimum particle number in mixed layer
```
This comes into play in `ocprod` to set the minimum particle number in the mixed layer
```
if ( k <= kmle ) then
   ocetra(i,j,k,inos) = MAX(nmldmin,ocetra(i,j,k,inos))
endif
```
and in the calculation of the `aggregate` and `dustagg` fields.

**Comment by TomasTorsvik** [comment]
One possible solution, if we want minimal changes to the behavior in iHAMOCC, would be to change `kmle` from an integer parameter to a 2D mask field, with values corresponding to the depth of the surface mixed layer. We could then make most of the changes very minimal by doing `do k=1,kmle(i,j)`, or something similar. Then we can also ensure that the new formulation can be backwards compatible with using isopycnic coordinates, i.e. set `kmle(i,j) = 2` everywhere.

**Comment by JorgSchwinger** [comment]
Hi Tomas
thanks for identifying all critical points related to kmld. I think your suggestion to use a spatially explicit kmle-array is good. We just have to discuss whether this can be passed by BLOM (mld is calculated somewhere in BLOM) or whether iHAMOCC calculates kmld (density is available, so calculating kmld according to some density criterion in mo_vgrid should be doable). 

 @matsbn, what do you think about passing an index field defining the depth of the mixed layer from BLOM to iHAMOCC?

**Comment by jmaerz** [comment]
Hi @JorgSchwinger and @TomasTorsvik , I was a bit reading on the preformed tracers (particularly oxygen-related), looked into the code and started wondering, why we update them through the full mixed layer depth and not only at surface? Particularly in deep mixed layers (polar & subpolar regions) this update thus includes already remineralization processes (when `kmle` > `kwrbioz`). With the new hybrid coordinates, this approach seems even more questionable to me. Do you have any thoughts on it? Any reasons for this approach (and not only updating surface values)?     

I just spoke briefly with Nadine about it and she mentioned some discussion with you, @JorgSchwinger earlier these days about the same issue. I am currently debating, whether it would make sense to introduce another preformed oxygen tracer (or all of the preformed tracers) in my simulations (only updating at the surface) which would allow to estimate the difference to what was done before.

**Comment by JorgSchwinger** [comment]
I think this is a question of how you define "preformed". My understanding was (is) that mixing inside the mixed layer is quick, so that it doesn't really make a difference. The bigger problem with the current implementation is how we initialize the preformed tracers: Currently it is initialized to zero (except in the mixed layer), so that it takes extremely long to get meaningful values in some places (deep North Pacific). It would probably be better to initialize to an estimate of preformed tracers using WOA/GLODAP - then it would still be off (=not an estimate according to our model) in less ventilated regions, but at least have a meaningful value there.

**Comment by jmaerz** [comment]
Most of it is oudated, closing the discussion.
